src/Network/WebSockets/TSSocketManager.m

```diff
@@ -218,17 +219,21 @@ - (void)processWebSocketRequestMessage:(WebSocketRequestMessage *)message {
     [self keepAliveBackground];
 
     if ([message.path isEqualToString:@"/api/v1/message"] && [message.verb isEqualToString:@"PUT"]) {
-        NSData *decryptedPayload =
-            [Cryptography decryptAppleMessagePayload:message.body withSignalingKey:TSStorageManager.signalingKey];
-
-        if (!decryptedPayload) {
-            DDLogWarn(@"Failed to decrypt incoming payload or bad HMAC");
-            return;
-        }
+        // SessionCipher state is mutated in a couple places, during decryption, in handleReceivedEnvelope.
+        // To ensure consistent state, that decryption mutation must occur on the same queue as encryption.
+        dispatch_async([OWSDispatch sessionCipher], ^{
+            NSData *decryptedPayload =
+                [Cryptography decryptAppleMessagePayload:message.body withSignalingKey:TSStorageManager.signalingKey];
+
+            if (!decryptedPayload) {
+                DDLogWarn(@"Failed to decrypt incoming payload or bad HMAC");
+                return;
+            }
 
-        OWSSignalServiceProtosEnvelope *envelope = [OWSSignalServiceProtosEnvelope parseFromData:decryptedPayload];
+            OWSSignalServiceProtosEnvelope *envelope = [OWSSignalServiceProtosEnvelope parseFromData:decryptedPayload];
 
-        [[TSMessagesManager sharedManager] handleReceivedEnvelope:envelope];
+            [[TSMessagesManager sharedManager] handleReceivedEnvelope:envelope];
```

Comment:
Per your other comment, this was an obvious place to dispatch `async` without further refactoring.

I could easily address what you've mentioned by getting ride of this dispatch_async call and  placing a couple more finely-grained `dispatch_sync` calls within this method.